### PR TITLE
Hash feature (compute and check hashes)

### PR DESCRIPTION
**Pull request by JeromeMartinez**
Still need some polish, but idea is there.

` --hash`
Compute or test the hash of files.
If input is raw content, do the hash and store hashes in RAWcooked metadata.
If an hash file is in the raw content, test that the hash file is valid i.e. that all hashes from hah file and hash of files match.
If input is a Matroska container and RAWcooked metadata contains hashes, decode input and check hashes (issue considered as decoding error).
If input is a Matroska container and with hash files in attachment, decode input and check hashes (issue considered as invalid data).

Slowness:
- As FFmpeg does the encoding, we need to read source content twice (1 by RAWcooked, 1 by FFmpeg)
- But if with ` --check`, we were previously already reading twice (1 by FFmpeg then 1 by RAWcooked checker part), now we compare hashes during the checker part so still twice (1 by RAWcooked, 1 by FFmpeg, source comparison is done based on hashes from 1st pass)

Supported hash file formats:
- md5sum (either by md5sum 1 file per directory, or 1 unique file by md5deep)
- md5sum --tags
- md5

### CODE REVIEW

File: Source/CLI/Help.cpp [hunk 1]
```diff
@@ -127,6 +100,56 @@ ReturnValue Help(const char* Name)
...
+    cout << "              If input is a Matroska container, decode input and check" << endl;
+    cout << "              that output is same as the decoded content." << endl;
+    cout << "       --no-check" << endl;
+    cout << "              Don't do checks (see above)." << endl;
+    cout << "              Is default (it may change in the future)" << endl;
+    cout << endl;
+    cout << "       --check-padding" << endl;
+    cout << "              Do costly (in terms of analysis duration or bytes read) checks." << endl;
+    cout << "              It is slower but guaranties reversability with e.g." << endl;
+    cout << "              DPX files with non zero padding bits." << endl;
```
Comment (retokromer) [inline Source/CLI/Help.cpp]: typo: reversability > reversibility

**Comment by JeromeMartinez** [comment]
Example with a [package having buggy md5 from different formats & different directories](https://github.com/MediaArea/RAWcooked-RegressionTestingFiles/tree/master/BuggyPackages/HashIssues) : 

```
rawcooked --hash --conch HashIssues
[...]
Invalid: Computed hashes are not same as hashes in hash files.
         HashIssues/image/0.txt
         HashIssues/image/000001.dpx
Invalid: files present but not listed in hash file.
         HashIssues/image/000002.dpx
Invalid: file names listed in hash file but not present.
         HashIssues/audio/audio2.wav
         HashIssues/image/000005.dpx
```

` --conch` is needed in order to not pollute the output with extra information not needed for everyone (without ` --conch`: only errors preventing encoding; with ` --conch`: errors preventing encoding + info about conformance of data, either DPX or md5sum)

Other tests are more complicated to produce, need to create intermediate files then make them corrupted.

**Comment by b00giehead** [comment]
I've just run a test here using your Beta version 18.10.1.20190329 on Windows 10. Had no problems at all. I was wondering if there's a way to check that all is well, and that the --hash function has worked at implanting a checksum in each file? Otherwise, do I assume that "Decoding was checked, no issues detected" is a sign of complete success?

**Comment by JeromeMartinez** [comment]
> "Decoding was checked, no issues detected"

I just launched a new build with a [fix about "decoding"](https://github.com/MediaArea/RAWcooked/commit/f81caa81838e8a48bc93657a1ff0c23f9f388a16), the behavior was not exactly the intended one if you use " --hash" while creating the file.

Build in progress (20190329-5 directory in [snapshots](https://mediaarea.net/download/snapshots/binary/rawcooked/)), intended behavior is when you check a compressed file:
- "Decoding was checked": no source info (either from RAWcooked hashes, or source content with " -o Directory" where "Directory" is the directory containing the directory you used as input during encoding, note that I plan to permit also to use the directory directly instead of of the directory containing the directory you used as input, so it is a bit more coherent between the command used for creating the file and the command for checking it), so we check only that decoding is fine (no decoding error).
- "Reversability was checked": source info is available (either RAWcooked hashes or source content with " -o"), we check decoding and that decoded files are exactly same as source files.

Then the outcome is indicated:
- ", no issue detected." --> the expected :) result.
- ", issues detected, see below." --> ho ho...

Does it seem logical?

I guess I need a write a bigger documentation here, because command to launch is easy, but explanations about what is done is less easy :(. I'll do it when features are more stable.

**Comment by JeromeMartinez** [comment]
Updated in https://github.com/MediaArea/RAWcooked/pull/323.

**Pull request merged**